storage/imx6ull-flashnor/flashnor-test.c

```diff
@@ -0,0 +1,157 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <meterfs.h>
+
+#include "storage.h"
+#include "flashnor-ecspi.h"
+
+
+static unsigned valcmp(uint8_t *buf, size_t len, uint8_t val)
+{
+	unsigned errors = 0;
+
+	for (size_t i = 0; i < len; i++) {
+		if (buf[i] != val)
+			errors++;
+	}
+
+	return errors;
+}
+
+
+void test_read_write(size_t flashsz, size_t sectorsz)
+{
+	int res = 0;
+	unsigned errors = 0;
+	uint8_t *pattern, *buf;
+
+	printf("%s: ", __func__);
+
+	pattern = malloc(sectorsz);
+	if (!pattern) {
+		printf("fail: malloc\n");
+		return;
+	}
+
+	buf = malloc(sectorsz);
+	if (!buf) {
+		printf("fail: malloc\n");
+		free(pattern);
+		return;
+	}
+
+	memset(pattern, 0x55, sectorsz);
+
+	/* Test only every sixteenth block */
+	for (unsigned int addr = 0; addr < flashsz; addr += sectorsz * 16) {
+		/* As meterfs is fixed right now in the driver implementation
+		   we have to erase sector before write, because of metadata written by meterfs */
+		if (flashnor_ecspiEraseSector(addr) < 0) {
+			printf("fail: erase sector");
+			res = -1;
+			break;
+		}
+
+		if (flashnor_ecspiWrite(addr, pattern, sectorsz) < 0) {
+			printf("fail: write sector at addr %u\n", addr);
+			res = -1;
+			break;
+		}
+
+		if (flashnor_ecspiRead(addr, buf, sectorsz) < 0) {
+			printf("fail: read sector at addr %u\n", addr);
+			res = -1;
+			break;
+		}
+
+		errors += valcmp(buf, sectorsz, pattern[0]);
+	}
+
+	if (res == 0) {
+		if (errors) {
+			printf("fail: read %u errors\n", errors);
+		}
+		else {
+			printf("ok\n");
+		}
+	}
+
+	free(pattern);
+	free(buf);
+}
+
+
+void test_erase(size_t flashsz, size_t sectorsz)
+{
+	int res = 0;
+	unsigned errors = 0;
+	uint8_t *buf;
+
+	printf("%s: ", __func__);
+
+	buf = malloc(sectorsz);
+	if (!buf) {
+		printf("fail: malloc\n");
+		return;
+	}
+
+	/* Clean sectors written by test_write_read */
+	for (unsigned int addr = 0; addr < flashsz; addr += sectorsz * 16) {
+		if (flashnor_ecspiEraseSector(addr) < 0) {
+			printf("fail: erase sector");
+			res = -1;
+			break;
+		}
+
+		if (flashnor_ecspiRead(addr, buf, sectorsz) < 0) {
+			printf("fail: erase sector read at addr %u\n", addr);
+			res = -1;
+			break;
+		}
+
+		errors += valcmp(buf, sectorsz, 0xff);
+	}
+
+	if (res == 0) {
+		if (errors) {
+			printf("fail: read %u errors\n", errors);
+		}
+		else {
+			printf("ok\n");
+		}
+	}
+
+	free(buf);
+}
+
+
+int main(int argc, char *argv[])
+{
+	int err, c;
+	meterfs_ctx_t *ctx; /* We use meterfs context because (for now) it's fixed in the driver */
+	storage_t dev = { 0 };
+	unsigned int ecspi_no = 3;
+
+	while ((c = getopt(argc, argv, "e:")) != -1) {
+		switch (c) {
+			case 'e':
+				ecspi_no = strtoul(optarg, NULL, 0);
+				break;
+		}
+	}
+
+	if ((err = flashnor_ecspiInit(ecspi_no, &dev)) < 0) {
+		printf("fail: failed to initialize device (err: %d)", err);
+		return 1;
+	}
+
+	ctx = (meterfs_ctx_t *)dev.ctx;
+
+	test_read_write(ctx->sz, ctx->sectorsz);
+	test_erase(ctx->sz, ctx->sectorsz);
```

Comment:
No need for this variable imho
```suggestion
	test_read_write(dev.ctx->sz, ctx->sectorsz);
	test_erase(dev.ctx->sz, ctx->sectorsz);
```